### PR TITLE
schema: Use generated dashboard model in frontend

### PR DESCRIPTION
**Pull request by sdboyer**
**What this PR does / why we need it**:

Now that #54816 is in, we're _actually finally ready to start using the generated dashboard types in the frontend omg omg omg ❤️ 🎉 🎊 🎉 ❤️_

I figured i'd open this PR that just swaps in the type to get the ball rolling. Simply doing so produces almost 200 errors to be fixed :)

Maybe we end up figuring out a way to chunk it down, and we can close this PR. But i figured this was the easiest way for everyone to be able to see the TODO list.

### CODE REVIEW

**Comment by sdboyer** [comment]
Just updated this to make it a `Partial<Dashboard>`, at least initially. This reduces the set of errors to deal with down to 121 from 174 - slightly more tractable :)

With this change, on a quick scan, it looks like the only non-test file with errors is `DashboardModel.ts` itself. Almost every other test issue appears to be certain required fields being absent from panels. Seems like something that the exported `defaultPanel` const could probably help mitigate?

File: pkg/coremodel/dashboard/coremodel.cue
Comment (ryantxu) [inline pkg/coremodel/dashboard/coremodel.cue:28]: oooh this is a weird on.  Looks like it is specific to dashboards supplied by plugins https://github.com/grafana/grafana/blob/main/public/app/types/plugins.ts#L15

**Comment by sdboyer** [comment]
Apologies for the disruption here - with the merge of the kind framework in #56492, there were a number of conflicts to resolve as "coremodel" is now "core kind" and the dashboard schema now lives at `kinds/structured/dashboard/dashboard_kind.cue`.

i took the liberty of merging and resolving the conflicts in a way that i _think_ should basically make this a no-op for you all - just editing the schema at a different path :)

File: public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx
Comment (polibb) [inline public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx:168]: Heyo 👾 After doing a mob session on fixing this error (`dashData` cannot be unknown, as it is passed to `DashboardModel` constructor which expects a `Dashboard`), we realised that there is an issue with [the getSaveModelClone() of DashboardModel](https://github.com/grafana/grafana/blob/use-veneer-dashboard-model/public/app/features/dashboard/state/DashboardModel.ts#L223). It returns a DashboardModel _type_, but in reality it does not return a new instance of the `class DashboardModel`, but a plain object with the properties in a Dashboard. Changing the return value to `Dashboard` creates much more errors everywhere else. We are wondering which way is best to go from here.

The `DashboardModel` has a lot more required fields than the `Dashboard` interface created by Cue does. We are wondering what the constraints are here about the json of a dashboard in the frontend vs. in the backend? @torkelo maybe you can answer this or point us in the right direction? Is the frontend supposed to be much more constrained than the backend when creating a new dashboard, or can the Cue schema of the dashboard have much more required fields set in itself?

Comment (sdboyer) [reply public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx:168]: i tried changing the return type of `getSaveModelClone` to the generated `Dashboard` type, and looked over a bunch of the errors. It seems like there are probably a few techniques that need to be applied.

But, before we get into those - would it be reasonable to defer those fixes to a follow-up PR? This is such a large task, it seems to me that the more we can break it down into stages, the better.

Comment (sdboyer) [reply public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx:168]: > The `DashboardModel` has a lot more required fields than the `Dashboard` interface created by Cue does.

We've [discussed this a bit in slack](https://raintank-corp.slack.com/archives/C03BH67DS2J/p1668528266330109) already, and its part of a larger, tricky topic - defaults and field optionality.

The key thing we need to start getting at is what it means for there to be a default value, vs. an optional field. I talked about that a bit in [this CUE issue](https://github.com/cue-lang/cue/discussions/1951#discussioncomment-3969260). 

But, concretely -

> or can the Cue schema of the dashboard have much more required fields set in itself?

In general, my guess is that it should be fine to make more fields required in the schema, WITH a default value.

> Is the frontend supposed to be much more constrained than the backend when creating a new dashboard

The main audience here to consider is probably more the as-code use cases, rather than the backend. Outside of public dashboards, there's little 'backend' code that makes a dashboard. When we think about that audience in the context of optional/default fields, it's pretty similar to what we think about in the frontend - what's the minimum the user _must_ state in an object in order to be a basically valid dashboard?

Comment (torkelo) [reply public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx:168]: @polibb yea, the getSaveModelClone returns the persisted state DTO so not DashboardModel

>  Is the frontend supposed to be much more constrained than the backend when creating a new dashboard

No, DashboardModel constructor is super forgiving, nothing is required as it applies defaults

File: public/app/features/dashboard/state/DashboardModel.ts
Comment (torkelo) [inline public/app/features/dashboard/state/DashboardModel.ts:128]:
```suggestion
  constructor(data: Partial<Dashboard>, meta?: DashboardMeta, private getVariablesFromState: GetVariables = getVariablesByKey) {    
```

Can't we change this constructor to take in a partial Dashboard model (like before)? Then we don't need all these calls to  createDashboardModelFixture({}) that to me obscures a bit what is going on. 

@joshhunt @ivanortegaalba

**Review by torkelo — COMMENTED** [review]
@sdboyer @ivanortegaalba @joshhunt 

I feel like having the schema types but named so plainly "Dashboard", "Panel", "VariableModel" is a bit confusing. The current type for dashboard was named DashboardDataDTO. 

would it make sense to suffix some of these types to make it more clear they are schema for the persisted/write (http) state? 

* Dashboard -> DashboardDTO 
* Panel > PanelDTO 
etc

But suffixing all might be messy, or suffixing them on import?

File: public/app/features/dashboard/state/DashboardModel.ts
Comment (joshhunt) [reply public/app/features/dashboard/state/DashboardModel.ts:128]: What is the type that comes over the wire? `Dashboard` or `Partial<Dashboard>`?

**Comment by sdboyer** [comment]
> suffixing them on import?

tl;dr - this seems like a fine convention, at least for short term.

---

> would it make sense to suffix some of these types to make it more clear they are schema for the persisted/write (http) state?

It seems almost certain to me that we're going to end up needing a suffix, or some similar pattern. But as you note,

> suffixing all might be messy

And that's true even if we're only considering this one case. But it seems unlikely this would be the only case. I suspect we're going to need schema variants - e.g., different representations of a schema for use different contexts. For example, we might need one that's all-optional for PATCH-like semantics. How do we deal with naming there? Do we generate `DashboardPatch`? `DashboardDTOPatch`?

Not having _any_ suffix on generated types feels equivalent to not taking a stance on this variants question. Not taking a stance until we're ready feels preferable to me. Doubly so because i think the "DTO" paradigm isn't really applicable to how k8s objects work, and that's the system we're most closely modeling after.

**Comment by sdboyer** [comment]
That said, this variants thing is organically coming up more and more. i'll write up an issue about it tomorrow so that we have a place to track it.

File: public/app/features/dashboard/state/DashboardModel.ts
Comment (torkelo) [reply public/app/features/dashboard/state/DashboardModel.ts:128]: @joshhunt Partial, as this also used in import where this ctor accept any old version, and them adds defaults + migration

Comment (joshhunt) [reply public/app/features/dashboard/state/DashboardModel.ts:128]: Yeah, that's the part im unsure about - old versions are not simply partial, but a completely different type The point is that migration takes a completely incompatible type and turns it into a Dashboard.

Comment (torkelo) [reply public/app/features/dashboard/state/DashboardModel.ts:128]: because the migration and defaults are applied inside the constructor (and not in the backend), it does accept anything (even an empty object).

Comment (torkelo) [reply public/app/features/dashboard/state/DashboardModel.ts:128]: So unless we start applying defaults and running migrations outside the constructor the type of that param is never really "correct", Partial made it a lot easier to use from unit tests etc, and forced it to handle dashboard json models with missing values

**Review by polibb — COMMENTED** [review]
@ivanortegaalba you rockstar! Awesome work on getting the context about everything dashboard-related and implementing it here. I'm looking forward to fixing all the rest self-contained needs of the model and schema 🙌.

There is a bug I find when trying to create a new panel in a new dashboard:

https://user-images.githubusercontent.com/13227501/207355315-70487d80-90ee-40e3-81fa-cb1524ae3ac5.mov

There is no bug when creating a new row or adding from library panels instead of creating a new panel.

**Review by polibb — COMMENTED** [review]
The bug mentioned above is fixed!

I don't understand the **breaking changes** because the reason seems to be removed `label`, `type`, and `name` of `VariableModel`, when they are not removed in fact. Do you have any idea what is happening here with this check @ivanortegaalba?

File: public/app/features/dashboard/state/DashboardMigrator.test.ts
Comment (polibb) [inline public/app/features/dashboard/state/DashboardMigrator.test.ts:69]: I took the liberty of moving this ts-expect-error here because I saw two errors from the typecheck:
```
public/app/features/dashboard/state/DashboardMigrator.test.ts:67:11 - error TS2578: Unused '@ts-expect-error' directive.

67           // @ts-expect-error
             ~~~~~~~~~~~~~~~~~~~

public/app/features/dashboard/state/DashboardMigrator.test.ts:70:13 - error TS2322: Type 'boolean' is not assignable to type '{ show: boolean; sort?: string | undefined; sortDesc?: boolean | undefined; }'.

70             legend: true,
               ~~~~~~
```

but now I sometimes see a fully successful pass of `yarn typecheck` and build the frontend successfully, and sometimes see some errors... I hope everything is alright, but please check one last time before you merge.

Comment (ivanortegaalba) [reply public/app/features/dashboard/state/DashboardMigrator.test.ts:69]: I experienced this problem as well. it seems to be a problem of how TS updates the cached results. If you remove from your local `tsconfig.tsbuildinfo` you will see the type check success deterministically.

**Comment by ivanortegaalba** [comment]
> 

It is because now the `list` templating [variable is optional](https://github.com/grafana/grafana/pull/55769/files#diff-2e74e4698f8f7f62e1bb57f8d729e635a38ac219c73e1188776a878b4502c151R75), so we're "removing" it because it is optional now.

**Comment by ivanortegaalba** [comment]
Here is a PR to fix the issues with Enterprise https://github.com/grafana/grafana-enterprise/pull/4255

**Review by ivanortegaalba — APPROVED** [review]
After multiple changes, and manual testing of the modified runtime code.  This PR is ready! ✅ Waiting for a second approval and I'll merge

**Review by polibb — APPROVED** [review]
Ran all the checks and cannot find any manual bugs 🙌.

**Pull request merged**
